Pretty names: support names set via export or Variable constructor

anderspapitto opened the pull request:
Add an argument to the Variable constructor in python. When supplied,
this string will be used as the unique name of the operator, rather
than an integer taken from a global counter. When provided, these names
must be unique among all Variables in the graph.

ezyang commented on torch/csrc/jit/ir.h:
```diff
@@ -565,6 +573,12 @@ friend struct Node;
   }
 
+  // In some cases, the compiler is not able do distinguish between
+  // overloads of `create` function. This helps.
```
LOL, I guess this is because strings can default construct into ArrayRef. It might be worth rearranging the API in a bigger way to solve this; perhaps setting the pretty name via a setter exclusively.

ezyang commented on torch/csrc/jit/ir.h:
```diff
@@ -659,12 +673,20 @@ friend struct Node;
   type_(getInitialType(kind_)) {
+  if (pretty_name_ != "") {
+    if (graph_->unique_pretty_names.find(pretty_name_) != graph_->unique_pretty_names.end()) {
+      // TODO: how do I raise an error?
```
You can use JIT_ASSERT(0) (or, more preferably, `JIT_ASSERT(graph_->unique_pretty_names.find(pretty_name_) != graph_->unique_pretty_names.end())`) to signify that there has been a fatal invariant broken, indicating developer error. Although, I'm not sure we should apply that here, because pretty names come from Python world, so they can trigger this condition themselves. In that case, raising a `std::runtime_error` is more appropriate.

ezyang commented:
CI failure is legit; it's a Python 2/3 portability problem. `torch/csrc/utils/python_strings.h` will probably be of interest.

houseroad reviewed (commented):
Actually, this PR goes deeper than I thought. In my mind, the solution is just changing the input name in protobuf. Does not need to touch the tracer part. But this PR can give a pretty name to every variable node in the graph. I am thinking about keeping this pretty name, and add unique id after the pretty name, something like "x-7". So it won't conflict, and JIT_ASSERT will makes more sense. And you can expose an API to the user to let them just set the names of the inputs and outputs.

colesbury commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -376,6 +380,14 @@ int THPVariable_set_requires_grad(THPVariable *self, PyObject *obj)
+PyObject *THPVariable_get_pretty_name(THPVariable* self)
+{
+  return self->cdata.pretty_name() == ""
+    ? NULL
```
Do you want to return `None`? Use the `Py_RETURN_NONE` macro. Returning NULL usually signifies an error.

apaszke commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -154,10 +154,11 @@ PyObject *THPVariable_pynew(PyTypeObject *type, PyObject *args, PyObject *kwds)
-  const char *accepted_args[] = {"data", "requires_grad", "volatile", "_grad_fn", NULL};
-  if (!PyArg_ParseTupleAndKeywords(args, kwds, "|ObbO", (char**)accepted_args,
-      &data, &requires_grad, &is_volatile, &grad_fn))
+  const char *accepted_args[] = {"data", "requires_grad", "volatile", "_grad_fn", "pretty_name", NULL};
```
If it's going to become part of public API then please let's just call it "name"

apaszke commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -186,6 +187,8 @@ PyObject *THPVariable_pynew(PyTypeObject *type, PyObject *args, PyObject *kwds)
     var = make_variable(torch::createTensor(data), requires_grad, is_volatile);
   }
 
+  var.pretty_name() = std::string(pretty_name ? pretty_name : "");
```
How about `if (pretty_name) var.pretty_name() = pretty_name;`? Might be faster in case `pretty_name` is NULL (which is most of the time) because it won't have to run `operator=` and additional constructor.

apaszke commented on torch/csrc/jit/ir.h:
```diff
@@ -156,9 +157,15 @@ struct Node : public Attributes<Node> {
+      ? pretty_name_
+      : std::to_string(unique())
+      ;
+    return debugPrefix + ret;
```
As seen in here, there is already a `debugName()` filed in all nodes. You can just reuse this to store the "name" taken from a Variable, and all these JIT modifications won't be necessary. If the "debug" prefix is what's worrying then we can always make it a "name"

anderspapitto commented:
Addressed inline comments

- reused debugName (changing it to prettyName)
- Py_RETURN_NONE
- THPUtils_packString
- public api is "name"
- don't ever throw exceptions/errors. duplicate names are now allowed

I intend to separately add a test in https://github.com/ezyang/onnx-pytorch/blob/master/test/test_models.py

After offline discussion with @houseroad, I intend to send a separate PR with naming of inputs/outports specifically for the export() function.

anderspapitto commented:
Commit message now reads

```
Variable: add support for friendly names
Add a "name" named parameter to the Variable constructor in python. When
supplied, the unique name of the corresponding Node will be
"<name parameter>_<unique integer id>". Names are not required to be
unique across Variables, since the integer id suffix allows
disambiguating between them.
```

anderspapitto commented:
Actually, since there's a slight dependency, and as far as I know github's not good for dependencies between code reviews - here I add a separate mechanism for setting input and output names during model export.

The only intersection between the two features (export and Variable) is in the uniqueName() function

houseroad commented on torch/onnx/__init__.py:
```diff
@@ -40,7 +40,7 @@ def set_training(model, mode):
 
 
-def export(model, args, f, export_params=True, verbose=False, training=False):
+def export(model, args, f, input_names=[], output_names=[], export_params=True, verbose=False, training=False):
```
Although input-names and output_names are not modified in our case,
I would suggest don't init the parameters with a mutable default value.

apaszke replied:
Yes, it's better to allocate lists inside the function and make `None` the defaults. Also, it's a better idea to add arguments at the end. So far this call `torch.onnx.export(model, args, f, True)` was valid, but wouldn't be after this patch

anderspapitto replied:
yeah, I forgot about this particular python feature for a moment. I'll change them to None, and at the end

houseroad commented on torch/onnx/__init__.py:
```diff
@@ -72,10 +72,10 @@ def export(model, args, f, export_params=True, verbose=False, training=False):
 
 
-def _export(model, args, f, export_params=True, verbose=False, training=False):
+def _export(model, args, f, input_names=[], output_names=[], export_params=True, verbose=False, training=False):
```
Same here.

houseroad commented on torch/csrc/jit/ir.cpp:
```diff
@@ -531,4 +531,29 @@ void PythonOp::cloneFrom(Node * other_) {
   }
 }
 
+void renameInputsAndOutputs(const std::shared_ptr<Graph>& graph,
```
Why do we need a rename function? Why not move the logic into NameInputsAndOutputs

anderspapitto replied:
just in case I would have had to mess around with the includes of export.h

However I'll change it as you suggest (unless it's rendered moot by @apaszke's suggestion below)

houseroad reviewed (changes requested):
In general, the PR did two things:
1) Support name in the Variable constructor, so we can name the variable during the tracing. Later we may extend it.
2) Name the inputs and outputs, so we can control the input and output names. It's helpful to connect different models.

@anderspapitto Please add the tests into onnx_pytorch repo.

apaszke commented on torch/csrc/autograd/variable.h:
```diff
@@ -109,6 +112,8 @@ struct VariableImpl : public at::TensorImpl {
   int output_nr;
   PyObject *pyobj;  // weak reference
 
+  std::string pretty_name;
```
Can we just call it `name` everywhere in autograd for consistency between C++ and Python API?

anderspapitto replied:
Well, PythonOp and CppOp inherit from Node and they both already have a name() method with a different meaning (the name of the underlying operation), so I preferred to not clash with that. 

If you don't mind that clash, or have some other suggestion, I'm happy to change it.

apaszke replied:
I think `name_` should be safe, right? (and `name` in autograd, since there's no clash for that)

apaszke commented on torch/csrc/jit/ir.h:
```diff
@@ -111,7 +111,8 @@ struct Node : public Attributes<Node> {
   size_t stage_ = 0;           // 0-forward, 1-backward, 2-double-backward,...
-  std::string debug_name_;
+  std::string pretty_name_;
+  std::string overriding_unique_name_;
```
How about just `name_` and `unique_name_`?

anderspapitto replied:
about `name_`: see my reply above
about `unique_name_`: since there's already a function called `uniqueName()` which does something more complicated than `return unique_name_`, I preferred to have a name that implies that something more complicated is going on (because it is).

Happy to change either/both though if you prefer the shorter names.

apaszke replied:
I thought to have `unique_name_` make `uniqueName()` return just this if it's set, and do the more complicated string dance otherwise

apaszke commented on torch/csrc/jit/ir.h:
```diff
@@ -456,7 +463,9 @@ struct Node : public Attributes<Node> {
   virtual Node * allocNewInstance(Graph * g) {
-    return new Node(g, kind());
+    Node * ret = new Node(g, kind());
+    ret->setPrettyName(prettyName());
```
I think this isn't supposed to copy any node attributes. This function is needed only because you can't have virtual constructors in C++, so you need to use it to allocate a "clone" of a node that has appropriate runtime info and memory layout

anderspapitto replied:
hmm. Ok, I'll take a closer look here

apaszke commented on torch/csrc/jit/ir.h:
```diff
@@ -515,8 +524,10 @@ friend struct Node;
 
-  Node * addInput() {
-    return addInput(create(kParam));
+  Node * addInput(std::string name_="") {
```
nit: why the underscore in argument name? It makes it easier to confuse with object attributes

anderspapitto replied:
I think in some intermediate state of this diff this was a constructor argument, and that seems to be the convention there (e.g. see Node::Node()).

I'll change it.

apaszke commented on torch/csrc/jit/ir.cpp:
```diff
@@ -531,4 +531,29 @@ void PythonOp::cloneFrom(Node * other_) {
+  for (size_t i = 0; i < output_names.size(); i++) {
+    graph->outputs()[i]->setOverridingUniqueName(output_names[i]);
+  }
+}
```
Since all of this functionality is exposed in Python, wouldn't it be simpler to remove all those C++ helpers for renaming and write this code in Python as a helper to `_export`?

anderspapitto replied:
modifying `overriding_unique_name_` isn't exposed in Python except through this mechanism (though of course I can expose it directly).

I wasn't aware that direct access to graph->inputs() and graph->outputs() was exposed either. If I find where that functionality is I can try moving this to python code.

apaszke reviewed (commented):
Looks good. Some minor comments, mostly about naming (it could be simplified).

anderspapitto commented:
addressed most inline comments, except that I haven't tried moving the export-naming functionality wholesale into the python code yet.

pretty_name -> name (in autograd only)
overriding_unique_name -> unique_name
merge Rename... into Name...
name_ -> name

houseroad commented on torch/csrc/jit/ir.h:
```diff
@@ -112,7 +112,8 @@ struct Node : public Attributes<Node> {
   size_t unique_ = 0;          // unique id
   size_t stage_ = 0;           // 0-forward, 1-backward, 2-double-backward,...
-  std::string debug_name_;
+  std::string pretty_name_;
```
What is the purpose of the field, and current status.

houseroad commented on torch/csrc/jit/ir.h:
```diff
@@ -112,7 +112,8 @@ struct Node : public Attributes<Node> {
   size_t stage_ = 0;           // 0-forward, 1-backward, 2-double-backward,...
-  std::string debug_name_;
+  std::string pretty_name_;
+  std::string unique_name_;
```
Same here

houseroad reviewed (approved):
LGTM. Good job, Anders!

houseroad commented:
There are some conflicts, please rebase onto master. @anderspapitto

Commit:
Pretty names: support names set via export or Variable constructor

Add (fully opt-in) functionality to support setting pretty names for
nodes in the graph. In particular

- Variable now has a `name` parameter in the constructor
- export now has `input_names` and `export_names` parameters

Nodes that are not named via this mechanism continue to be named
internally with unique integers.

Names have a few rules.

- They must all be unique in the graph.
- They may not be integers (because of potential conflicts with
  internally generated names).

anderspapitto commented:
I put some more thought into this, and changed the interface again.

My primary reasoning is "this feature is a minor convenience, and it's not worth much complexity".

Storing both a pretty_name and a unique_name, automatically appending suffixes - these aren't worth the complexity (for example, a pretty_name_ set on node A with an automatic suffix appended could conflict with a unique_name_ set on node B).

Now there's one name stored, and it's stored exactly as you set it, and there's only two rules
- everything is unique
- names can't be integers

I'm open to more discussion and/or going back to how it was, though at this point I'd rather just drop the Variable(name=...) feature entirely than have it operate through a slightly-different mechanism from export() like it was doing before.

anderspapitto commented:
commit message now reads:

```
Add (fully opt-in) functionality to support setting pretty names for
nodes in the graph. In particular

- Variable now has a `name` parameter in the constructor
- export now has `input_names` and `export_names` parameters

Nodes that are not named via this mechanism continue to be named
internally with unique integers.

Names have a few rules.

- They must all be unique in the graph.
- They may not be integers (because of potential conflicts with
  internally generated names).
```

houseroad commented:
I would only keep one name field, and make sure all the names are unique. Two fields increases unnecessary complexity. If the name conflicts, raise the exception, the name is for debugging purpose. So the chance to get the conflicts should be pretty low.

anderspapitto commented:
@houseroad great, sounds like we agree. That's exactly what I did.

Looks like I broke a bunch of tests though, so I'll get those fixed.

anderspapitto commented:
test failure look spurious to me - it passes locally and is unrelated to the change

dzhulgakov reviewed (commented):
@ezyang @zdevito - can we merge it? It'd make tutorials nicer to look at

soumith commented:
@pytorchbot test this please

soumith commented:
thanks Anders!
